ocaml/typing/typemod.ml

```diff
@@ -1607,6 +1617,75 @@ let mksig desc env loc =
 
 (* let signature sg = List.map (fun item -> item.sig_type) sg *)
 
+(** When inferring a module type upon seeing a [Pmtd_underscore], we first apply a
+    substitution from the paths appearing in the signature to the paths appearing in
+    the structure. If after the substitution we encounter a [Path.t] not recognised by
+    the env of the struct, this means the module type being inferred has a symbol not yet
+    included in the struct, and so we should error out. This avoids having confusing
+    error messages later.
+
+    We do this check by checking if the head name (e.g [A] in [A.t]) doesn't appear
+    in the current structure but does appear in the signature. If it does appear in the
+    current structure, then all is well; if it doesn't appear in the signature, then it
+    was declared somewhere outside so it can't be unbound. This way, we avoid expensive
+    [Env.t] queries.*)
+let check_no_unbound_paths env loc sig_map str_map mty =
+  let check_error path in_sig in_str =
+    if in_sig && (not in_str) then
+      raise (Error (loc, env, Unbound_path_in_inferred_type path))
+  in
+  let iterator =
+    { Btype.type_iterators with
+      Btype.it_path = (fun path_kind path ->
+        match path with
+        | Pident id ->
+            let name = Ident.name id in
+            let (in_sig, in_str) =
+              begin match path_kind with
+                | Path_value ->
+                    Sig_map.has_value name sig_map,
+                    Sig_map.has_value name str_map
+                | Path_type ->
+                    Sig_map.has_type name sig_map,
+                    Sig_map.has_type name str_map
+                | Path_module ->
+                    Sig_map.has_module name sig_map,
+                    Sig_map.has_module name str_map
+                | Path_modtype ->
+                    Sig_map.has_module_type name sig_map,
+                    Sig_map.has_module_type name str_map
+                | Path_class ->
+                    Sig_map.has_class name sig_map,
+                    Sig_map.has_class name str_map
```

Comment:
Similarly, I suspect that module type can't refer to a class. Let's discuss this offline.